File: GitToolBox/src/main/java/zielu/gittoolbox/revision/RevisionInfo.java
package zielu.gittoolbox.revision;

import com.intellij.openapi.vcs.history.VcsRevisionNumber;
import java.time.ZonedDateTime;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public interface RevisionInfo {
  RevisionInfo EMPTY = new RevisionInfo() {
    private static final String EMPTY_TEXT = "EMPTY";

    @NotNull
    @Override
    public VcsRevisionNumber getRevisionNumber() {
      return VcsRevisionNumber.NULL;
    }

    @NotNull
    @Override
    public String getAuthor() {
      return EMPTY_TEXT;
    }

    @NotNull
    @Override
    public ZonedDateTime getDate() {
      return ZonedDateTime.now();
    }

    @Nullable
    @Override
    public String getSubject() {
      return null;
    }

    @Override
    public boolean isEmpty() {
      return true;
    }

    @Override
    public boolean isNotEmpty() {
      return false;
    }
  };

  @NotNull
  VcsRevisionNumber getRevisionNumber();

  @NotNull
  String getAuthor();

  @NotNull
  ZonedDateTime getDate();

  @Nullable
  String getSubject();

  boolean isEmpty();

  boolean isNotEmpty();
}

File: GitToolBox/src/main/kotlin/zielu/gittoolbox/revision/RevisionInfo.kt
package zielu.gittoolbox.revision

import com.intellij.openapi.vcs.history.VcsRevisionNumber
import java.time.ZonedDateTime

interface RevisionInfo {
  fun getRevisionNumber(): VcsRevisionNumber
  fun getAuthor(): String
  fun getSubject(): String?
  fun isEmpty(): Boolean
  fun isNotEmpty(): Boolean
  fun getAuthorEmail(): String?
  fun getDate(): ZonedDateTime

  companion object {
    @JvmField
    val NULL = object : RevisionInfo {
      override fun getRevisionNumber() = VcsRevisionNumber.NULL
      override fun getAuthor() = "EMPTY"
      override fun getAuthorEmail(): String? = null
      override fun getDate() = ZonedDateTime.now()
      override fun getSubject(): String? = null
      override fun isEmpty() = true
      override fun isNotEmpty() = false
    }
  }
}
